exp/operator/controllers/genericprovider_controller.go

```diff
@@ -92,14 +104,175 @@ func (r *GenericProviderReconciler) Reconcile(ctx context.Context, req reconcile
 	return r.reconcile(ctx, typedProvider, typedProviderList)
 }
 
+func (r *GenericProviderReconciler) secretReader(ctx context.Context, genericProvider genericprovider.GenericProvider) (configclient.Reader, error) {
+	mr := configclient.NewMemoryReader()
+
+	// TODO: maybe set a shorter default, so we don't block when installing cert-manager.
+	//mr.Set("cert-manager-timeout", "120s")
+
+	if genericProvider.GetSpec().SecretName != nil {
+		secret := &corev1.Secret{}
+		key := types.NamespacedName{Namespace: genericProvider.GetNamespace(), Name: *genericProvider.GetSpec().SecretName}
+		if err := r.Client.Get(ctx, key, secret); err != nil {
+			return nil, err
+		}
+		for k, v := range secret.Data {
+			mr.Set(k, string(v))
+		}
+	}
+
+	if genericProvider.GetSpec().FetchConfig != nil && genericProvider.GetSpec().FetchConfig.URL != nil {
+		mr.WithProvider(genericProvider.GetName(), util.ClusterctlProviderType(genericProvider), *genericProvider.GetSpec().FetchConfig.URL)
+	}
+
+	return mr, nil
+}
+
+func (r *GenericProviderReconciler) configmapRepository(ctx context.Context, genericProvider genericprovider.GenericProvider) (repository.Repository, error) {
+	// TODO get the configMap using spec.FetchConfig.Selector and populate an in-memory repository.
+	return nil, nil
+}
+
 func (r *GenericProviderReconciler) reconcile(ctx context.Context, genericProvider genericprovider.GenericProvider, genericProviderList genericprovider.GenericProviderList) (_ ctrl.Result, reterr error) {
 	// Run preflight checks to ensure that core provider can be installed properly
 	result, err := preflightChecks(ctx, r.Client, genericProvider, genericProviderList)
 	if err != nil || !result.IsZero() {
 		return result, err
 	}
 
-	return ctrl.Result{}, nil
+	reader, err := r.secretReader(ctx, genericProvider)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	cfg, err := configclient.New("", configclient.InjectReader(reader))
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	providerConfig, err := cfg.Providers().Get(genericProvider.GetName(), util.ClusterctlProviderType(genericProvider))
+	if err != nil {
+		conditions.Set(genericProvider, conditions.FalseCondition(
+			operatorv1.PreflightCheckCondition,
+			operatorv1.UnknownProviderReason,
+			v1alpha4.ConditionSeverityWarning,
+			fmt.Sprintf(unknownProviderMessage, genericProvider.GetName()),
+		))
+		return ctrl.Result{}, nil
+	}
+
+	clusterClient := cluster.New(cluster.Kubeconfig{}, cfg)
+
+	// ensure the custom resource definitions required by clusterctl are in place
+	if err := clusterClient.ProviderInventory().EnsureCustomResourceDefinitions(); err != nil {
```

Comment:
Is this a kind of preflight check for installing a provider? If so maybe there is a way to include it in `preflightChecks()`

Reply:
if they don't exist, it creates the CRDs. So not really a preflight check..